node/overseer/overseer-gen/proc-macro/src/impl_builder.rs

```diff
@@ -107,46 +62,234 @@ pub(crate) fn impl_builder(info: &OverseerInfo) -> proc_macro2::TokenStream {
 		})
 		.collect::<Vec<_>>();
 
-	let generics = quote! {
-		< S, #( #baggage_generic_ty, )* >
-	};
-	let where_clause = quote! {
+	// Helpers to use within quote! macros
+	let builder_where_clause = quote! {
 		where
-			S: #support_crate ::SpawnNamed,
+			#(
+				#subsystem_type : Subsystem<#subsystem_ctx_name< #consumes >, #error_ty>,
+			)*
+			S: #support_crate ::SpawnNamed + Send
 	};
-
 	let builder_generics = quote! {
-		<S, #( #baggage_generic_ty, )* #( #builder_generic_ty, )* >
+		S, #( #subsystem_type, )* #( #baggage_generic_ty, )*
 	};
 
-	// all subsystems must have the same context
-	// even if the overseer does not impose such a limit.
-	let builder_additional_generics = quote! {
-		<#( #builder_generic_ty, )* >
-	};
+	let placeholder_type =
+		create_all_placeholder_idents(subsystem_name.as_slice(), baggage_name.as_slice());
+	let field_name = subsystem_name.iter().chain(baggage_name.iter()).collect::<Vec<_>>();
+	let field_type = subsystem_type
+		.iter()
+		.chain(info.baggage().iter().filter_map(|bag| bag.field_ty.get_ident()))
+		.collect::<Vec<_>>();
 
-	let consumes = &info.consumes();
+	// Setters logic
 
-	let subsyste_ctx_name =
-		Ident::new(&(overseer_name.to_string() + "SubsystemContext"), overseer_name.span());
+	// For each setter we need to leave the remaining fields untouched and
+	// remove the field that we are fixing in this setter
+	// For subsystems we also need _with and replace_ setters
+	let subsystem_specific_setters = info
+		.subsystems()
+		.iter()
+		.filter(|ssf| !ssf.wip)
+		.enumerate()
+		.map(|(idx, ssf)| {
+			let fname = &ssf.name;
+			let ftype = &ssf.generic;
+			let subsystem_consumes = &ssf.consumes;
+			// Remove placeholder at specific field position
+			let impl_generics = &(placeholder_type[..idx]
+				.iter()
+				.chain(placeholder_type[idx + 1..].iter())
+				.collect::<Vec<_>>());
+			let fname_with = format_ident!("{}_with", fname);
+			let fname_replace = format_ident!("replace_{}", fname);
+
+			let uninit_generics = produce_setter_generic_permutation(
+				ftype,
+				idx,
+				placeholder_type.as_slice(),
+				&Ident::new("OverseerFieldUninit", Span::call_site()),
+			);
+			let return_type_generics = produce_setter_generic_permutation(
+				ftype,
+				idx,
+				placeholder_type.as_slice(),
+				&Ident::new("OverseerFieldInit", Span::call_site()),
+			);
+			let other_field_name =
+				field_name[..idx].iter().chain(field_name[idx + 1..].iter()).collect::<Vec<_>>();
+			let replace_modified_generics = return_type_generics
+				.iter()
+				.enumerate()
+				.map(|(other_idx, generic)| {
+					if other_idx == idx {
+						quote!(OverseerFieldInit<NEW>)
+					} else {
+						generic.clone()
+					}
+				})
+				.collect::<Vec<_>>();
+
+			quote! {
+				#[allow(non_camel_case_types)]
```

Comment:
We should be able to manage without this, right?

Reply:
No, removing this attribute produces a warning like the following:

```
warning: type parameter `__plinkos` should have an upper camel case name
  --> node/overseer/overseer-gen/examples/dummy.rs:96:2
   |
96 |     plinkos: GoblinTower,
   |     ^^^^^^^ help: convert the identifier to upper camel case: `Plinkos`
   |
   = note: `#[warn(non_camel_case_types)]` on by default
```

In fact, whenever we use any combination of the placeholder types, except full initialized or full uninitialized, we need to add this override.

This particular place is transformed to the following output in `cargo expand --example dummy`:

```
impl<S, AwesomeSubSys, __plinkos, __i_like_pi, __i_like_generic>
    XxxBuilder<S, OverseerFieldUninit<AwesomeSubSys>, __plinkos, __i_like_pi, __i_like_generic>
where
    AwesomeSubSys: Subsystem<XxxSubsystemContext<MsgStrukt>, Yikes>,
    GoblinTower: Subsystem<XxxSubsystemContext<Plinko>, Yikes>,
    S: crate::SpawnNamed + Send,
```

Reply:
Because the type parameter should be `GoblinTower` (the generic).

Reply:
`__i_like_pi` should also not become a generic in the builder, the type is fixed in the declaration. That's the difference between baggage and subsystem. Baggage items must be explicitly declared as generic or as a type, any `#[subsystem]` annotation will make it an _implicit_ (= generated) generic where the user just gives the name (for orientation, this usually matches the subsystem implementation struct name), which we should then continue using as generic.

Reply:
The idea was to create a generic for each field, so a builder has a generic parameter (placeholder) for each field. Then we specialize builder impl for each field being set or unset and we use generics for this type of specializations where needed (e.g. for subsystems or generic baggage). The builder struct itself looks like the following one:

```
pub struct XxxBuilder<S, __sub0, __plinkos, __i_like_pi, __i_like_generic> {
    sub0: __sub0,
    plinkos: __plinkos,
    i_like_pi: __i_like_pi,
    i_like_generic: __i_like_generic,
    spawner: ::std::option::Option<S>,
}
```

So no original generics are preserved in the initial structure.